Preference Save actions - open console and set L&F

DanVanAtta opened the pull request:
### Code Update Overview
- Added a save action callback that can be used to perform an action when settings are changed
- Updated enum configuration to provide a callback and update L&F or open the console window when 'console is always open' is set to true.

### Functional changes

#### Console Window:

Now opens when 'always show console window is toggled from false->true and save button is clicked.
![show_console](https://user-images.githubusercontent.com/12397753/36636109-66798b90-1974-11e8-9524-dab85184de6b.png)

#### Look and Feel (L&F) Setting:

L&F  is now applied to the UI immediately on save instead of application start. The settings text is updated to request a game re-start after saving. In this way the new L&F can be previewed as early as the save confirmation window.

![on_save](https://user-images.githubusercontent.com/12397753/36636123-a73ea64c-1974-11e8-85f9-e9dcbb76534c.png)

An easier preview of L&F was requested in: https://forums.triplea-game.org/topic/537/preview-for-look-and-feel-setting

DanVanAtta commented:
Testing this did not turn out like I had pictured. I am open to good suggestions for how to test; otherwise at this point I'd like to submit these updates having done by-hand testing.

ssoloff commented on game-core/src/main/java/games/strategy/triplea/settings/ClientSetting.java:
```diff
@@ -92,7 +95,11 @@
 
-  SHOW_CONSOLE_ALWAYS(false),
+  SHOW_CONSOLE_ALWAYS(false, saveValue -> {
+    if (saveValue.equals("true")) {
```
Nit: Since we're relying on the `String`-to-`boolean` conversions elsewhere, we should probably be consistent and do the same thing here (and guard against an NPE since `ClientSetting#setValue()` allows `null` values).

ssoloff replied:
Sorry, forgot the code example:

```java
if (String.valueOf(true).equals(saveValue)) {
```

ssoloff commented on game-core/src/main/java/games/strategy/triplea/settings/ClientSetting.java:
```diff
@@ -218,6 +238,7 @@ public boolean isSet() {
 
   @Override
   public void save(final String newValue) {
+    onSaveAction.accept(newValue);
```
One way to test this is to reverse the dependency.  Instead of `ClientSetting` knowing what the save action is, allow it to be specified by an external client.  That's probably better for decoupling in the long run.  For example, I'm not sure if `ClientSetting` should know about `ErrorConsole`, much less contain behavior manipulating it.

Anyway, here's a rough cut at allowing multiple save listeners per setting (note that it leaves the current save actions in place and doesn't try to move them outside this type, but that would be the goal in the final realization):

```diff
diff --git a/game-core/src/main/java/games/strategy/triplea/settings/ClientSetting.java b/game-core/src/main/java/games/strategy/triplea/settings/ClientSetting.java
index 7c71bc164..03fc02039 100644
--- a/game-core/src/main/java/games/strategy/triplea/settings/ClientSetting.java
+++ b/game-core/src/main/java/games/strategy/triplea/settings/ClientSetting.java
@@ -3,6 +3,8 @@ package games.strategy.triplea.settings;
 import static com.google.common.base.Preconditions.checkNotNull;
 
 import java.io.File;
+import java.util.Collection;
+import java.util.concurrent.CopyOnWriteArrayList;
 import java.util.concurrent.atomic.AtomicReference;
 import java.util.function.Consumer;
 import java.util.prefs.BackingStoreException;
@@ -139,13 +141,13 @@ public enum ClientSetting implements GameSetting {
 
   private static final AtomicReference<Preferences> preferencesRef = new AtomicReference<>();
 
-  private final Consumer<String> onSaveAction;
+  private final Collection<Consumer<String>> saveListeners = new CopyOnWriteArrayList<>();
 
   public final String defaultValue;
 
   ClientSetting(final String defaultValue, final Consumer<String> onSaveAction) {
     this.defaultValue = defaultValue;
-    this.onSaveAction = onSaveAction;
+    saveListeners.add(onSaveAction);
   }
 
   ClientSetting(final boolean defaultVAlue, final Consumer<String> onSaveAction) {
@@ -238,7 +240,7 @@ public enum ClientSetting implements GameSetting {
 
   @Override
   public void save(final String newValue) {
-    onSaveAction.accept(newValue);
+    saveListeners.forEach(listener -> listener.accept(newValue));
     getPreferences().put(name(), newValue);
   }
 
@@ -264,4 +266,14 @@ public enum ClientSetting implements GameSetting {
   public void resetAndFlush() {
     saveAndFlush(defaultValue);
   }
+
+  @Override
+  public void addSaveListener(final Consumer<String> listener) {
+    saveListeners.add(listener);
+  }
+
+  @Override
+  public void removeSaveListener(final Consumer<String> listener) {
+    saveListeners.remove(listener);
+  }
 }
diff --git a/game-core/src/main/java/games/strategy/triplea/settings/GameSetting.java b/game-core/src/main/java/games/strategy/triplea/settings/GameSetting.java
index cac7322f2..163e1cf5c 100644
--- a/game-core/src/main/java/games/strategy/triplea/settings/GameSetting.java
+++ b/game-core/src/main/java/games/strategy/triplea/settings/GameSetting.java
@@ -1,5 +1,7 @@
 package games.strategy.triplea.settings;
 
+import java.util.function.Consumer;
+
 /**
  * Basic save/read API for a 'game setting'. Generally these will be values the user can modify over the
  * course of the game, either directly or indirectly. For example, default window size may saved here,
@@ -38,4 +40,8 @@ public interface GameSetting {
   }
 
   void resetAndFlush();
+
+  void addSaveListener(Consumer<String> listener);
+
+  void removeSaveListener(Consumer<String> listener);
 }
diff --git a/game-core/src/test/java/games/strategy/triplea/settings/ClientSettingTest.java b/game-core/src/test/java/games/strategy/triplea/settings/ClientSettingTest.java
new file mode 100644
index 000000000..3229bda90
--- /dev/null
+++ b/game-core/src/test/java/games/strategy/triplea/settings/ClientSettingTest.java
@@ -0,0 +1,38 @@
+package games.strategy.triplea.settings;
+
+import static org.mockito.Mockito.verify;
+
+import java.util.function.Consumer;
+
+import org.junit.jupiter.api.AfterEach;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.extension.ExtendWith;
+import org.mockito.Mock;
+
+import com.example.mockito.MockitoExtension;
+
+@ExtendWith(MockitoExtension.class)
+public final class ClientSettingTest extends AbstractClientSettingTestCase {
+  @Mock
+  private Consumer<String> saveListener;
+
+  @BeforeEach
+  public void addSaveListener() {
+    ClientSetting.TEST_SETTING.addSaveListener(saveListener);
+  }
+
+  @AfterEach
+  public void removeSaveListener() {
+    ClientSetting.TEST_SETTING.removeSaveListener(saveListener);
+  }
+
+  @Test
+  public void shouldInvokeSaveListenersWhenNewValueSaved() {
+    final String newValue = "newValue";
+
+    ClientSetting.TEST_SETTING.save(newValue);
+
+    verify(saveListener).accept(newValue);
+  }
+}
```

DanVanAtta replied:
Interesting. Looking at this closer, wouldn't every `Consumer<String>` be invoked every time?

DanVanAtta replied:
I am considering having for example ErrorConsole register with the right client setting. I'm not sure about this; i like the tight cohesion of having all related configurations together in one place. Ideally you could add an entire setting with a new enum entry. 

I'm going to attempt to restructure things so that ClientSetting configuration is offloaded to a builder that we could test.

ssoloff replied:
Not sure if I understand the question.  Every `Consumer<String>` for the specific `ClientSetting` whose value was changed would be invoked, yes.  That's no different than what's happening currently except that each `ClientSetting` can now have more than one listener.  I used a collection of listeners for each setting instead of one per setting because, if you move the listener implementations outside this type, you really don't know how many clients may register for notification of a setting change.

DanVanAtta replied:
My misunderstanding on that point, it is 1:1 by enum value and not all enums.

@ssoloff this is a good suggestion.

The one thing I have been wondering about is initialization references. `ErrorConsole` should work as we init it explicitly on startup. On the other hand other components that could have save actions, but are not yet initialized, could be missed.

Perhaps we can resolve this be extracting all such classes/initialization to a common spot and then initialize it all together in GameRunner.

DanVanAtta replied:
I gave the suggestion a try, the initialization ordering issue turned out to be a problem.

I'm also hesitant to let 'save actions' be registered from anywhere in the code base. This seems to be poor cohesion. It would probably be okay if all actions were localized to the same package, but otherwise if it really grew you would have a hard time tracking down every save action listener; would be hard to know what would happen if we did a reset to factory default.

DanVanAtta replied:
> I gave the suggestion a try, the initialization ordering issue turned out to be a problem.

Derp, need to click save. We probably should make it so settings just update as they are changed instead of needing the 'save'

ssoloff commented:
Note that when I switch my L&F from SubstanceTwilight to Nimbus, I get the following stack trace (on Fedora 27):

```
java.lang.NullPointerException
	at org.pushingpixels.substance.internal.utils.SubstanceColorSchemeUtilities.getAlpha(SubstanceColorSchemeUtilities.java:347)
	at org.pushingpixels.substance.internal.utils.icon.ArrowButtonTransitionAwareIcon.getIconToPaint(ArrowButtonTransitionAwareIcon.java:163)
	at org.pushingpixels.substance.internal.utils.icon.ArrowButtonTransitionAwareIcon.paintIcon(ArrowButtonTransitionAwareIcon.java:135)
	at org.pushingpixels.substance.internal.ui.SubstanceComboBoxUI.paint(SubstanceComboBoxUI.java:529)
	at org.pushingpixels.substance.internal.ui.SubstanceComboBoxUI.update(SubstanceComboBoxUI.java:711)
	at javax.swing.JComponent.paintComponent(JComponent.java:780)
	at javax.swing.JComponent.paint(JComponent.java:1056)
	at javax.swing.JComponent.paintChildren(JComponent.java:889)
	at javax.swing.JComponent.paint(JComponent.java:1065)
	at javax.swing.JComponent.paintChildren(JComponent.java:889)
	at javax.swing.JComponent.paint(JComponent.java:1065)
	at javax.swing.JViewport.paint(JViewport.java:728)
	at javax.swing.JComponent.paintToOffscreen(JComponent.java:5210)
	at javax.swing.BufferStrategyPaintManager.paint(BufferStrategyPaintManager.java:290)
	at javax.swing.RepaintManager.paint(RepaintManager.java:1272)
	at javax.swing.JComponent._paintImmediately(JComponent.java:5158)
	at javax.swing.JComponent.paintImmediately(JComponent.java:4969)
	at javax.swing.RepaintManager$4.run(RepaintManager.java:831)
	at javax.swing.RepaintManager$4.run(RepaintManager.java:814)
	at java.security.AccessController.doPrivileged(Native Method)
	at java.security.ProtectionDomain$JavaSecurityAccessImpl.doIntersectionPrivilege(ProtectionDomain.java:80)
	at javax.swing.RepaintManager.paintDirtyRegions(RepaintManager.java:814)
	at javax.swing.RepaintManager.paintDirtyRegions(RepaintManager.java:789)
	at javax.swing.RepaintManager.prePaintDirtyRegions(RepaintManager.java:738)
	at javax.swing.RepaintManager.access$1200(RepaintManager.java:64)
	at javax.swing.RepaintManager$ProcessingRunnable.run(RepaintManager.java:1732)
	at java.awt.event.InvocationEvent.dispatch(InvocationEvent.java:311)
	at java.awt.EventQueue.dispatchEventImpl(EventQueue.java:756)
	at java.awt.EventQueue.access$500(EventQueue.java:97)
	at java.awt.EventQueue$3.run(EventQueue.java:709)
	at java.awt.EventQueue$3.run(EventQueue.java:703)
	at java.security.AccessController.doPrivileged(Native Method)
	at java.security.ProtectionDomain$JavaSecurityAccessImpl.doIntersectionPrivilege(ProtectionDomain.java:80)
	at java.awt.EventQueue.dispatchEvent(EventQueue.java:726)
	at java.awt.EventDispatchThread.pumpOneEventForFilters(EventDispatchThread.java:201)
	at java.awt.EventDispatchThread.pumpEventsForFilter(EventDispatchThread.java:116)
	at java.awt.EventDispatchThread.pumpEventsForFilter(EventDispatchThread.java:109)
	at java.awt.WaitDispatchSupport$2.run(WaitDispatchSupport.java:184)
	at java.awt.WaitDispatchSupport$4.run(WaitDispatchSupport.java:229)
	at java.awt.WaitDispatchSupport$4.run(WaitDispatchSupport.java:227)
	at java.security.AccessController.doPrivileged(Native Method)
	at java.awt.WaitDispatchSupport.enter(WaitDispatchSupport.java:227)
	at java.awt.Dialog.show(Dialog.java:1084)
	at javax.swing.JOptionPane.showOptionDialog(JOptionPane.java:869)
	at javax.swing.JOptionPane.showMessageDialog(JOptionPane.java:666)
	at javax.swing.JOptionPane.showMessageDialog(JOptionPane.java:637)
	at games.strategy.triplea.settings.SettingsWindow.lambda$9(SettingsWindow.java:172)
	at java.util.Optional.ifPresent(Optional.java:159)
	at games.strategy.triplea.settings.SettingsWindow.saveSettings(SettingsWindow.java:170)
	at swinglib.JButtonBuilder.lambda$0(JButtonBuilder.java:50)
	at javax.swing.AbstractButton.fireActionPerformed(AbstractButton.java:2022)
	at javax.swing.AbstractButton$Handler.actionPerformed(AbstractButton.java:2348)
	at javax.swing.DefaultButtonModel.fireActionPerformed(DefaultButtonModel.java:402)
	at javax.swing.DefaultButtonModel.setPressed(DefaultButtonModel.java:259)
	at javax.swing.plaf.basic.BasicButtonListener.mouseReleased(BasicButtonListener.java:252)
	at org.pushingpixels.substance.internal.utils.RolloverButtonListener.mouseReleased(RolloverButtonListener.java:126)
	at java.awt.Component.processMouseEvent(Component.java:6533)
	at javax.swing.JComponent.processMouseEvent(JComponent.java:3324)
	at java.awt.Component.processEvent(Component.java:6298)
	at java.awt.Container.processEvent(Container.java:2236)
	at java.awt.Component.dispatchEventImpl(Component.java:4889)
	at java.awt.Container.dispatchEventImpl(Container.java:2294)
	at java.awt.Component.dispatchEvent(Component.java:4711)
	at java.awt.LightweightDispatcher.retargetMouseEvent(Container.java:4888)
	at java.awt.LightweightDispatcher.processMouseEvent(Container.java:4525)
	at java.awt.LightweightDispatcher.dispatchEvent(Container.java:4466)
	at java.awt.Container.dispatchEventImpl(Container.java:2280)
	at java.awt.Window.dispatchEventImpl(Window.java:2746)
	at java.awt.Component.dispatchEvent(Component.java:4711)
	at java.awt.EventQueue.dispatchEventImpl(EventQueue.java:758)
	at java.awt.EventQueue.access$500(EventQueue.java:97)
	at java.awt.EventQueue$3.run(EventQueue.java:709)
	at java.awt.EventQueue$3.run(EventQueue.java:703)
	at java.security.AccessController.doPrivileged(Native Method)
	at java.security.ProtectionDomain$JavaSecurityAccessImpl.doIntersectionPrivilege(ProtectionDomain.java:80)
	at java.security.ProtectionDomain$JavaSecurityAccessImpl.doIntersectionPrivilege(ProtectionDomain.java:90)
	at java.awt.EventQueue$4.run(EventQueue.java:731)
	at java.awt.EventQueue$4.run(EventQueue.java:729)
	at java.security.AccessController.doPrivileged(Native Method)
	at java.security.ProtectionDomain$JavaSecurityAccessImpl.doIntersectionPrivilege(ProtectionDomain.java:80)
	at java.awt.EventQueue.dispatchEvent(EventQueue.java:728)
	at java.awt.EventDispatchThread.pumpOneEventForFilters(EventDispatchThread.java:201)
	at java.awt.EventDispatchThread.pumpEventsForFilter(EventDispatchThread.java:116)
	at java.awt.EventDispatchThread.pumpEventsForHierarchy(EventDispatchThread.java:105)
	at java.awt.EventDispatchThread.pumpEvents(EventDispatchThread.java:101)
	at java.awt.EventDispatchThread.pumpEvents(EventDispatchThread.java:93)
	at java.awt.EventDispatchThread.run(EventDispatchThread.java:82)
```

DanVanAtta commented:
> Note that when I switch my L&F from SubstanceTwilight to Nimbus, I get the following stack trace (on Fedora 27):

That stack trace is in substance?  I think I ran across that problem before, basically some L&F's are broken (we used to have a poison pill bug about this that woudl crash startup).

One problem with making preview easy is we'll probably hit that problem more frequently. On plus side ti is mitigated so the bad L&F will not continue being default. We perhaps should try to remove these bad L&F's with a hardcoded list. To some extent though I'd like to follow up with that in another PR, if I understand right it would be an existing problem.

ssoloff commented:
@DanVanAtta The above stack trace occurs whenever I switch from _any_ Substance L&F to _any_ non-Substance L&F.  It basically renders the application useless.  An empty error message dialog keeps popping up, and many UI controls (the ones still rendered using the Substance L&F) stop responding to input.

Yes, the root cause is most likely in Substance itself.  However, I don't think we should leave the user in a state where they basically have to kill the app after they make a setting change.

@RoiEXLab @ron-murhammer Could one of you please check out this branch and try the following:

1. Change the L&F to any Substance (if needed)
1. Change the L&F to any non-Substance (e.g. Nimbus)

Do you see an error dialog, and does the app essentially become non-functional?

If this only happens for me, I have no problem merging this PR.  However, if not, I don't think we should knowingly introduce such a regression for the benefit of "previewing" a L&F.

ron-murhammer commented:
@ssoloff After you restart TripleA, does the problem go away?

ssoloff commented:
@ron-murhammer Yes, but in order to restart you have to close all windows with the system close button (X) or nuke the process.  The normal UI buttons are unresponsive.

ron-murhammer commented:
@ssoloff I think I'm seeing the same problem. If I try switching to Nimbus I get a bunch of error popups and it locks up:
![image](https://user-images.githubusercontent.com/1427689/36649884-d45f3172-1a65-11e8-8a72-12e42105e5b1.png)

ssoloff commented:
@ron-murhammer Yup, that's it.  IMO, that's pretty bad, and we probably shouldn't include the "L&F preview" feature until it's resolved.  But maybe I'm making this out to be a bigger problem than it is?

ron-murhammer commented:
@DanVanAtta So my long term vision was to either remove L&F options or very limited set. Hoping if @RoiEXLab UI revamp pans out then we are more image based so you see less of just L&F stuff. Then picking maybe just 1 dark and 1 light L&F and customizing around those.

ssoloff commented:
> Is it an option to make user confirm that they need to restart tripleA after restarting? IE: they click save, we'll prompt a "are you sure? this setting will make TripleA unstable and you will need to restart the app"

@DanVanAtta I don't think we should let the user get into an unusable state even if they agree to it.

I'd say just leave things the way they are, but if you want to enhance the UX for the L&F setting a bit, you could always display a dialog that says something like, "You must restart TripleA for the new setting to take effect."  I think we say that in the setting descriptions that require such a restart, but most users probably don't read the description, or that particular line of text could out-of-sight below the scroll pane viewport.

DanVanAtta commented:
Now updated:
- removed L&F update
- took the suggestion for how to restructure to test. I'm concerned about save action cohesion; but at moment prefer to have the testing.

ssoloff commented on game-core/src/main/java/games/strategy/triplea/settings/ClientSetting.java:
```diff
@@ -131,17 +132,19 @@
 
   private static final AtomicReference<Preferences> preferencesRef = new AtomicReference<>();
-
+  private final List<Consumer<String>> onSaveAction = new ArrayList<>();
```
Would recommend using a `CopyOnWriteArrayList` here.  I don't think you can guarantee that `addSaveListener()` and `save()` will be called on the same thread, so this collection needs to be thread-safe.  In addition, if someone ever calls `addSaveListener()` concurrently while this collection is being iterated over in `save()`, it might trigger a `ConcurrentModificationException`.

---

Nit: Might want to pluralize the field name, too, since it is now a collection.

DanVanAtta replied:
>  I don't think you can guarantee that addSaveListener() and save() will be called on the same thread

No, but at same time `addSaveListener()` should always be some sort of initialization call. Notice that I did not add the 'removeListener' method as suggested, the saveAction should be immutable.

I noticed also that we call 'save' on app launch, this seems wrong, I opened a bug ticket: https://github.com/triplea-game/triplea/issues/3223

------

So, `CopyOnWriteArrayList` I don't think is quite the right answer as it implies there is a race condition. The fix for that is to not have the race. Save is expected to be driven via user action, so we should in effect be able to control the threading because of that contract.

ssoloff replied:
> No, but at same time addSaveListener() should always be some sort of initialization call. 

Then maybe we should enforce no `addSaveListener()` calls after the first `save()` call?  Otherwise, that seems like a buried assumption someone may not realize six months down the road.

> So, `CopyOnWriteArrayList` I don't think is quite the right answer as it implies there is a race condition.

The race isn't the only concern.  There's also safely publishing values between threads.  Without some sort of synchronization, there's no guarantee when the publication will happen other than "eventually."

DanVanAtta replied:
> Then maybe we should enforce no addSaveListener() calls after the first save() call?   Otherwise, that seems like a buried assumption someone may not realize six months down the road.

Interesting considerations. I don't think I want to lock that assumption in just yet though.

DanVanAtta commented:
<wip>
- fix bad merge
- make variable plural

DanVanAtta commented on game-core/src/main/java/games/strategy/engine/framework/lookandfeel/LookAndFeel.java:
```diff
@@ -41,11 +42,24 @@
+          JOptionPane.WARNING_MESSAGE);
+    });
+  }
+
```
note, this is an extra warning message asking user to restart:
![warning_message](https://user-images.githubusercontent.com/12397753/36942077-6fa1d820-1f1f-11e8-9e8a-0af40f9e3fe9.png)

After closing you get the standard property changed pop-up.

DanVanAtta commented on game-core/src/main/java/games/strategy/triplea/TripleA.java:
```diff
@@ -82,7 +83,9 @@ public void shutDown() {
-      game.removeDisplay(display);
+      if (game != null) {
+        game.removeDisplay(display);
+      }
```
this is a NPE fix while I was nearby, not strictly related.

DanVanAtta commented on game-core/src/main/java/games/strategy/triplea/TripleA.java:
```diff
@@ -116,6 +119,7 @@ public void startGame(final IGame game, final Set<IGamePlayer> players, final bo
     } else {
       SwingAction.invokeAndWait(() -> {
         final TripleAFrame frame = new TripleAFrame(game, localPlayers);
+        LookAndFeelSwingFrameListener.register(frame);
```
This static call is to the helper class above. It's important we update the UI of any component that is open after changing the UI manager. So any frame we create we need to register so it can be updated. We get automatic listener removal as well when the frame is closed.

DanVanAtta commented on game-core/src/main/java/games/strategy/triplea/settings/ClientSetting.java:
```diff
@@ -211,13 +205,26 @@ public static void flush() {
+        .orElseThrow(() -> new IllegalStateException("ClientSetting framework has not been initialized. "
+            + "Did you forget to call ClientSetting#initialize() in production code "
+            + "or ClientSetting#setPreferences() in test code?"));
+  }
```
this is a cleanup of method on line 180, and also moved it to be below its first line of usage.

DanVanAtta commented on game-core/src/main/java/games/strategy/triplea/settings/ClientSetting.java:
```diff
@@ -163,36 +170,23 @@
     preferencesRef.set(null);
   }
 
-  private static Preferences getPreferences() {
```
this method was moved lower

DanVanAtta commented on game-core/src/main/java/games/strategy/triplea/settings/ClientSetting.java:
```diff
@@ -163,36 +170,23 @@
-  }
-
-  @VisibleForTesting
-  static void setPreferences(final Preferences preferences) {
```
this method was moved lower. Since it is visible for testing, we can do parameter null check at point of caller, so the  precondition check was moved to 'initialize'

DanVanAtta commented:
@ron-murhammer @ssoloff this is updated; please have a look

### Updates:
- I managed to get L&F update working without crash.
- Added an extra warning dialog when updating UI theme to restart all instances (even though the game still does work well; it's other instances which are still running that could I found could potentially see problems (which is in theory an existing problem))
- Fixed UI theme checkbox to be pre-selected to the current setting

### Testing
- Tested with Mac+linux
- checked normal flow of main menu to settings and updated
- checked case where download maps window is open
- checked case of updating UI theme when the game is actually runnng (setting is available via file menu)
- :warning:  check into case where another instance was running and launched a lobby bot game after updating (we see the L&F update). I did run into an error here though; I was not able to always repro, also the game would lock-up but the lock-up would eventually go away.

ssoloff commented on game-core/src/main/java/games/strategy/triplea/settings/ClientSetting.java:
```diff
@@ -163,36 +170,23 @@
-
-    preferencesRef.set(preferences);
+    setPreferences(
+        Preconditions.checkNotNull(Preferences.userNodeForPackage(ClientSetting.class)));
```
Note that `Preferences#userNodeForPackage()` is documented to never return `null`, so `checkNotNull()` here is redundant.

DanVanAtta replied:
This was moved from `setPreferences`, this is the one non-test location that calls that method. That would mean the 'checkNotNull'  was redundant in that method as well.

DanVanAtta replied:
removed in: 25ea84e2d

ssoloff commented on game-core/src/main/java/games/strategy/triplea/settings/ClientSetting.java:
```diff
@@ -163,36 +170,23 @@
-    return preferences;
+  @Override
+  public void addSaveListener(final Consumer<String> saveListener) {
+    onSaveActions.add(saveListener);
```
We might want to consider a `checkNotNull()` precondition here simply because this is a public API, and adding a `null` listener will trigger an NPE at some time in the future (during iteration) rather than when it was added.

DanVanAtta replied:
On the one hand it would mean we added new code that passed null, on the other hand we can get slightly better error detection. A null value would not work well for listener removal, and doing that instead of a no-op consumer is really quite wrong. Hopefully we'd see it during CR...

DanVanAtta replied:
At the cost of a bit of code clutter, the error detection is good to have: 817d7f121
Some point down the road we won't be passing null values anymore, we won't have to dedicate code to it anymore as any nulls would be all class-internal, any public facing nulls are either optional or calls the right API.

ssoloff commented on game-core/src/main/java/games/strategy/triplea/settings/ClientSetting.java:
```diff
@@ -211,13 +205,26 @@ public static void flush() {
 
   @Override
   public void save(final String newValue) {
+    onSaveActions.forEach(saveAction -> saveAction.accept(newValue));
```
Also a location where a `checkNotNull()` precondition may be appropriate, just so a `null` doesn't escape into the listeners.  Although, it will still "almost" fail-fast because the call to `Preferences#put()` on the next line will throw an NPE if `newValue` is `null`.

DanVanAtta replied:
Filechooser has an example where an empty value is valid input. Also maybe for clearing out a value. We could change API to: `Consumer<Optional<String>>`

But, the original ClientSettings API used String and made decision that `Optional.empty` was equivalent to both `null` and the empty string. 

We can convert nulls to empty string here to help guarantee this. Ideally, it would be enforced upstream.

DanVanAtta replied:
null to empty string guard added: 07347f9b7

ssoloff replied:
Just to make sure I understand your intention here...  You want to allow `null` values and just silently convert them to empty strings?  If so, line 228 will still trigger an NPE if `newValue` is `null`:

```java
getPreferences().put(name(), newValue);
```

DanVanAtta replied:
That would be a problem, I updated the code to do a 'remove' if `newValue` is null: cda8f428a

The original contract of `ClientSettings` is that saved values are always non-null. FileChooser I think broke that potentially, but it might also never return null. It may be the case that we never see null passed to this method. If it is not, I do not want to refactor that now to fix it. So to maintain that original contract, that `null` and `empty string` are equivalent, they are both empty option preference, we can convert null to empty string here to at least guarantee it from this point on.

The alternative is to convert to a `Consumer<Optional<String>>`, but original clientSettings already had that consideration, I recall I chose to go for empty string to be treated the same as an Optional.empty, so no Optionals are required.

ssoloff commented on game-core/src/main/java/games/strategy/engine/framework/lookandfeel/LookAndFeel.java:
```diff
@@ -41,11 +42,24 @@
+      SettingsWindow.updateLookAndFeel();
+      JOptionPane.showMessageDialog(
+          null,
+          "WARNING: look and feel changes can cause instability. Please restart all running TripleA instances",
```
Nit: Missing a period at the end of the message.  Only flagging because the intention appears to have been to use two sentences due to the period after _instability_.

DanVanAtta replied:
FWIW - Intent, was for minimal punctuation. Given we have two sentences, without a line break we need a period there : )

We could probably use a style guide; whether to prefer the final period or not. I'm also noticing that we have all our 'yes' buttons on the right. This is starting to seem wrong to me, for better or worse it does seem like default, positive menu selection is on left. 

Getting back to this topic, I tweaked this message a bit, no need to say 'WARNING' when in a warning dialog already, and I broke up to two line and made the trailing period consistent:

91ea813a4

DanVanAtta replied:
Getting back to this topic, I tweaked this message a bit, no need to say 'WARNING' when in a warning dialog already, and I broke up to two line and made the trailing period consistent:

91ea813a4

DanVanAtta commented:
Let's get this merged for wider testing. Thanks for feedback/review. @panther2 heads up that this update makes L&F easier to change. It should be okay to not restart TripleA after changing it, but we could see crashes. Any such places we can/should fix.
